### ISSUE TITLE
cli/sql: error following successful non-row-returning query is not reported

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#16519]
With `cockroach sql`

```
> SELECT 1; SELECT INVALID;
+---+
| 1 |
+---+
| 1 |
+---+
(1 row)
```

(no error reported)

edit: the cause of the problem is in `lib/pq` - filed as https://github.com/lib/pq/issues/633

**Comment by knz** [comment]
cc @mjibson @jordanlewis

**Comment by jordanlewis** [comment]
This is present in 1.0 and master. For the record, two consecutive statements without errors work correctly in both branches:

```
root@:26257/> SELECT 1; SELECT 2;
+---+
| 1 |
+---+
| 1 |
+---+
(1 row)
+---+
| 2 |
+---+
| 2 |
+---+
(1 row)
```

Also, as a related comment, Postgres executes all statements when separated by semicolons even if there's an error before the last statement:

```
test=# select b; select 1;
ERROR:  42703: column "b" does not exist
LINE 1: select b;
               ^
LOCATION:  errorMissingColumn, parse_relation.c:3183
 ?column?
----------
        1
(1 row)
```

For us, however, an error precludes execution of the rest of the statements:

```
root@:26257/> SELECT f; SELECT 2;
pq: column name "f" not found
root@:26257/>
```

**Comment by knz** [comment]
I think avoiding to run subsequent statements is a good default. If user demand comes we could add it as an option. However the error should be reported nonetheless.

**Comment by knz** [comment]
The problem is upstream in `lib/pq`, filed as https://github.com/lib/pq/issues/633

cc @mjibson

**Comment by mjibson** [comment]
As discussed in the lib/pq issue, I currently think the best solution to this problem is to have the cli split up these statements at semicolons and send them one at a time. This would give us more correct error tracking, but it would also break some auto-retry stuff in cockroach, so it's unclear which is better.

**Comment by mjibson** [comment]
As an update: I'm not convinced the original bug posted here can't be fixed with lib/pq as is. lib/pq does return an error for this, just not when expected. So I think cockroach is doing something wrong.

**Comment by andreimatei** [comment]
This issue seems to still coming up. Do you think you'll have any time to look into it in the foreseeable future?

**Comment by mjibson** [comment]
Yes, after feature freeze when I have time to work on bugs.

**Comment by mjibson** [comment]
I can no longer reproduce this.

```
root@:26257/> SELECT 1; SELECT invalid;
pq: column name "invalid" not found
root@:26257/> SELECT 1; SELECT 1/0;
pq: division by zero
```

So fixed maybe? On both master and v1.1-alpha.20170810.

**Comment by knz** [comment]
It's not *entirely* fixed, look at this:

```
root@:26257/t> SELECT 1; SELECT 2; select foo;
+---+
| 1 |
+---+
| 1 |
+---+
(1 row)

Time: 997.995µs

pq: column name "foo" not found
```

It seems that the results from the statement immediately preceding the error is swallowed.

**Comment by mjibson** [comment]
Yes, which is https://github.com/lib/pq/issues/633 and has a known but involved fix. It feels unlikely to make it into 1.1 since it'd be too risky merge at this point. But at least the real bad bug as described in the topic is fixed.

**Comment by mjibson** [comment]
Moving to 1.2 because the full fix for this is a significant refactor that won't happen in 1.1.

**Comment by rytaft** [comment]
I found this issue again during QA for another feature. As described above, if there are multiple statements on the same line, and one causes an error, there is no indication that the previous statement was successful.
```
root@:26257/> CREATE TABLE t.b (i int, b bool); INSERT INTO b VALUES (2, true), (3, true), (4, false);
pq: relation "b" does not exist
```
^ table `t.b` was actually created, but the second statement failed since `b` was unqualified.

**Comment by jordanlewis** [comment]
@rytaft's repro still works. I'll move it to the issue subject, since the original repro doesn't work anymore. I feel like this is probably quite fixable now, since it works fine with *queries* that return rows, just not statements that don't.

**Comment by jordanlewis** [comment]
Actually, since the original issue is totally fixed, I'll open a new issue for the remaining problem.

**Comment by jordanlewis** [comment]
#55234

**Comment by knz** [comment]
FYI the original issue at top is fixed at least in v21.1. Either due to a change in lib/pq or inside crdb.
